src/search/pdbs/pattern_collection_generator_multiple.cc

```diff
@@ -0,0 +1,311 @@
+#include "pattern_collection_generator_multiple.h"
+
+#include "pattern_database.h"
+#include "utils.h"
+
+#include "../option_parser.h"
+#include "../plugin.h"
+
+#include "../utils/countdown_timer.h"
+#include "../utils/logging.h"
+#include "../utils/markup.h"
+#include "../utils/rng.h"
+#include "../utils/rng_options.h"
+
+#include <vector>
+
+using namespace std;
+
+namespace pdbs {
+PatternCollectionGeneratorMultiple::PatternCollectionGeneratorMultiple(
+    options::Options &opts)
+    : max_pdb_size(opts.get<int>("max_pdb_size")),
+      pattern_generation_max_time(opts.get<double>("pattern_generation_max_time")),
+      total_max_time(opts.get<double>("total_max_time")),
+      stagnation_limit(opts.get<double>("stagnation_limit")),
+      blacklist_trigger_percentage(opts.get<double>("blacklist_trigger_percentage")),
+      enable_blacklist_on_stagnation(opts.get<bool>("enable_blacklist_on_stagnation")),
+      verbosity(opts.get<utils::Verbosity>("verbosity")),
+      rng(utils::parse_rng_from_options(opts)),
+      random_seed(opts.get<int>("random_seed")),
+      remaining_collection_size(opts.get<int>("max_collection_size")),
+      blacklisting(false),
+      stagnation_start_time(-1) {
+}
+
+void PatternCollectionGeneratorMultiple::check_blacklist_trigger_timer(
+    double blacklisting_start_time, const utils::CountdownTimer &timer) {
+    // Check if blacklisting should be started.
+    if (!blacklisting && timer.get_elapsed_time() > blacklisting_start_time) {
+        blacklisting = true;
+        // Also reset stagnation timer in case it was already set.
+        stagnation_start_time = -1;
+        if (verbosity >= utils::Verbosity::NORMAL) {
+            utils::g_log << "given percentage of total time limit "
+                         << "exhausted; enabling blacklisting." << endl;
+        }
+    }
+}
+
+unordered_set<int> PatternCollectionGeneratorMultiple::get_blacklisted_variables(
+    vector<int> &non_goal_variables) {
+    unordered_set<int> blacklisted_variables;
+    if (blacklisting && !non_goal_variables.empty()) {
+        /*
+          Randomize the number of non-goal variables for blacklisting.
+          We want to choose at least 1 non-goal variable and up to the
+          entire set of non-goal variables.
+        */
+        int blacklist_size = (*rng)(non_goal_variables.size());
```

Comment:
I agree that this is ugly to read. You could ask in Discord if there is any reason for this interface and if otherwise, it is worth opening an issue for this.

Reply:
I put this question to discord and close it here.